tools/base/abstract.py

```diff
@@ -0,0 +1,97 @@
+from abc import ABCMeta
+
+
+class Implementer(type):
+    """Metaclass for implementers of an Abstract interface
+
+    Any `Abstraction` classes that are listed in `__implements__` for the
+    class are added as bases (ie added to the class's inheritance).
+
+    Any docs for methods are copied from the interface method to the
+    implementation if the implementation method has no docs of its own.
+
+    For example:
+
+    ```
+
+    from tools.base.abstract import Abstraction, Implementer
+
+
+    class AFoo(metaclass=Abstraction):
+
+        @abstractmethod
+        def do_something(self):
+            \"""Do something\"""
+            raise NotImplementedError
+
+    class Foo(metaclass=Implementer)
+        __implements__ (AFoo, )
```

Comment:
This bit seems a bit funky; is there any way to make `AFoo` part of the inheritance hierarchy?

Reply:
it is added as part of the inheritance (bases) under the hood

mostly i was trying to find a way to avoid inheritance, at least explicitly, to try and make it more like a composable/polymorphic system

with ABCs you dont have to inherit directly from an ABC, you can do `ABC.register(MyImpl)` but its not *exactly* the same as `MyImpl(ABC)` just very close

initially i diidnt do this because i couldnt get mypy to recognize the `virtual types`. In the end i needed to hack mypy anyway, and using inheritance has the advantage that the descendant class is checked (by python) as implementing the ABC interface, so i kept it this way

Reply:
figuring out whether to use `ABC.register()` or as i have done using a metaclass to add the inheritance bases is the bit im most uncertain of in this implementation

if we go with `.register` that makes the ABCs more like pure interfaces - ie you cant do `super().some_iface_method()` to make use of the abstract implementation (although i havent added any abstract implementations, partly for this reason)

apart from superclassing, the advantage of using inheritance is that the `abstractmethods` are checked when the class is instantiated. If any are missing a `TypeError` is thrown

im ambivalent about this becuase it seems to add type-checking at runtime, which is not how python generally works

also, if we dont use the inheritance pattern, i can probs make the metaclass do this checking anyway

Reply:
in summary, i have implemented the way i have, so that it will work with genuinely abstract classes, even if im only using it for interfaces

Reply:
also worth mentioning is that the expected use is with a decorator, ie

```python
@abstract.implementer(AFoo)
class Foo:
   ...

```

so the `__implements__` etc would be abstracted out

Reply:
OK, with the decorator this is very clean, thanks.